Fix some encoding issues (mostly on Windows)

PhysSong opened the pull request:
This pull request fixes several encoding issue. Every issue except STK one are Windows-specific.
- ZynAddSubFX setting loading/saving
- VST DLL loading
- VST setting loading/saving
- Sample file decoding
- MIDI import
- STK rawwave path (incomplete on Windows due to upstream issue)

Issues remaining on Windows:
- .pat file loading
- .gig file loading
- STK rawwave path (which `toLocal8Bit` fails to handle)

I'm planning to track remaining issues in #1191.

Fixes #3855, Fixes #2662, and almost #1191.

zonkmachine commented:
> Every issue except STK one are Windows-specific

I replicated issue #2662 on LinuxMint/Qt4 . It's indeed fixed by this PR.

Ready to merge?

Wallacoloo commented on plugins/zynaddsubfx/zynaddsubfx/src/Misc/QtXmlWrapper.cpp:
```diff
@@ -193,9 +202,18 @@ int QtXmlWrapper::dosavefile(const char *filename,
+#else
+    int fd = dup(f.handle());
+#endif
+    f.close();
```
I see you're using this to open files by their descriptor instead of the filename. Is it `gzopen` that's giving you problems, `fopen`, or both? Because if it's only `gzopen` that's broken, it might be possible to just always `fopen`, and then call `gzdopen(fileno(file), options)` if compression is enabled, rather than doing this trickery with duplicating the file handles and such.

PhysSong replied:
Both are broken, because C library only supports characters in local code page.

Wallacoloo replied:
Oh, oops. I had typed out that comment and after digging deeper to better understand the problem I meant to remove it before posting the review >.<

Yeah, I agree with your conclusion here. It's not pretty, but I don't see any better solution.

Wallacoloo commented on plugins/zynaddsubfx/zynaddsubfx/src/Misc/QtXmlWrapper.cpp:
```diff
@@ -313,7 +331,16 @@ int QtXmlWrapper::loadXMLfile(const std::string &filename)
+#else
+    int fd = dup(f.handle());
+#endif
+    f.close();
```
I wonder if you could offload this logic to a helper function,
```
static int fdopenFromQString(QString filename, QIODevice::OpenMode mode)
```
to avoid duplicating the body in multiple functins, and then leave a comment near that function explaining why the function is necessary (to most readers, this code will look silly and they may be tempted to simplify it back to the way it was before).

PhysSong replied:
Good suggestion!

Wallacoloo commented on src/core/DrumSynth.cpp:
```diff
@@ -162,17 +164,20 @@ float DrumSynth::waveform(float ph, int form)
+    QFile f(file);
+    f.open(QIODevice::ReadOnly);
+    QByteArray dat = f.readAll().constData();
+    is.str(string(dat.constData(), dat.size()));
```
The casual reader will see this and think "we're reading the entire file into a string, and then operating on it only as a stream? Why not directly open the file with `ifstream`?" and will be tempted to change this back to how it was previously. I think there should be some [short] comment explaining why we _must_ open the file with `QFile`, possibly linking back to this github PR.

Wallacoloo commented on src/core/SampleBuffer.cpp:
```diff
@@ -416,7 +412,10 @@ f_cnt_t SampleBuffer::decodeSampleSF( const char * _f,
 
-	if( ( snd_file = sf_open( _f, SFM_READ, &sf_info ) ) != NULL )
+
+	QFile f(_f);
```
I'm thinking these should all be commented with something along the lines of
```
// Open the file with QFile and pass the descriptor to sf_open_fd instead of directly calling sf_open
// for better filename encoding support. See https://github.com/LMMS/lmms/pull/4401
```
so that people know not to revert this change.

PhysSong replied:
Good point.

PhysSong commented on plugins/vst_base/RemoteVstPlugin.cpp:
```diff
@@ -119,6 +119,29 @@ RemoteVstPlugin * __plugin = NULL;
+static std::wstring toWString(std::string s)
+{
+	/*
+	QString qstr = QString::fromUtf8(s.data(), int(s.size()));
```
Beacause it requires including Qt headers(on Windows). If that don't matter (we already do that), I can switch to QString is needed.

Wallacoloo replied:
Ah, I see. @lukas-w has offered me advice on where it's okay to use Qt stuff in the past. I think things like QString, at least, are fine everywhere, but hopefully he'll chime in here to confirm that this is OK if he's not too busy.

PhysSong replied:
To be more clear, `RemoteVstPlugin` use Qt libraries for IPC on Windows. So I think that would be fine.

lukas-w replied:
@PhysSong is right, this is not a good place to use Qt because that would add MinGW Qt as a dependency on Linux builds. We can however use C++11 functions to convert to UTF-16:
```c++
#include <locale>
#include <codecvt>
```
```c++
std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>> converter;
return converter.from_bytes(s);
```

This does the same thing, just in fewer lines. I think we already use C++11 for `RemoteVstPlugin` on `stable-1.2`, but I didn't check.

PhysSong replied:
> because that would add MinGW Qt as a dependency on Linux builds.

It seems like we can use host version Qt, but I don't think it's a good idea because it adds 32bit Qt as a dependency on 64bit Linux.

> We can however use C++11 functions to convert to UTF-16:

That doesn't work with our old MinGW because of missing header files(it should work with newest `g++-mingw-w64`). Those functions are deprecated in C++17, too.

PhysSong commented on plugins/vst_base/RemoteVstPlugin.cpp:
```diff
@@ -119,6 +119,29 @@ RemoteVstPlugin * __plugin = NULL;
+#define F_OPEN(a, b) _wfopen(toWString(a).c_str(), L##b)
+#else
+#define F_OPEN(a, b) fopen((a).c_str(), b)
+#endif
```
Forgot to commit that. Thank you!

PhysSong replied:
Addressed.

PhysSong commented on plugins/MidiImport/MidiImport.cpp:
```diff
@@ -281,6 +284,11 @@ bool MidiImport::readSMF( TrackContainer* tc )
+	if (!file.open(QIODevice::ReadOnly))
+	{
+		return false;
+	}
```
I wrote like this because `m_outputFile` is a private member of `AudioFileDevice`. I'll add a method to work around.

Wallacoloo commented:
@PhysSong Thanks for the improvements!

I'm marking this as "approved" now, since you've addressed any concerns I had. But it looks like Lukas wanted to review this too, so you should probably wait until that before merging :tada:

lukas-w commented on plugins/stk/mallets/mallets.cpp:
```diff
@@ -620,7 +620,12 @@ malletsSynth::malletsSynth( const StkFloat _pitch,
+#else
+						.toUtf8().constData() );
+#endif
+
```
I think we can (should?) use `toLocal8Bit` on Unix as well. It should be equivalent to `toUtf8` on most Linux and macOS systems. Correct me if I'm wrong.

lukas-w commented on plugins/zynaddsubfx/zynaddsubfx/src/Misc/QtXmlWrapper.cpp:
```diff
@@ -193,9 +216,11 @@ int QtXmlWrapper::dosavefile(const char *filename,
                            int compression,
                            const char *xmldata) const
 {
+   int fd = fdopenFromUtf8(filename, QIODevice::WriteOnly);
```
Why not use the same method here that's used in `RemoteVstPlugin.cpp`? We can move the `toWString` method and the `F_OPEN` defines to a common header and include that in both files.

PhysSong replied:
I think `F_OPEN` macro is an ugly workaround for `fopen`. I can do that if needed, but I don't like the macro approach.

lukas-w replied:
@PhysSong You can make it a function instead.

PhysSong replied:
Something tricky about that: `"r"` vs `L"r"` or something like that. That's why I chose macro approach.

lukas-w replied:
@PhysSong Sounds like you need to convert the mode string as well. So something like this:

```c++
std::FILE* openFile(std::string filename, std::string mode)
{
#ifdef LMMS_BUILD_WIN32
	return _wfopen(toWString(filename), toWString(mode));
#else
	return fopen(filename.data(), mode.data());
#endif
}
```

PhysSong replied:
Looks fine for me, except missing `data()` for WIN32.

lukas-w replied:
Can you elaborate on that?

PhysSong replied:
I mean `return _wfopen(toWString(filename).data(), toWString(mode).data());`.

lukas-w replied:
Of course, sorry. I thought you were telling me `data()` isn't available for WIN32 and started getting angry at MinGW. :)

PhysSong replied:
I don't really know which approach will be better here. Both approach seem to have not much overheads. Qt provides a nice API for this, but `_wfopen` looks somewhat straightforward.
Alternatively, `CreateFileW` + [`_open_osfhandle`](https://msdn.microsoft.com/en-us/library/bdts1c9x.aspx)(what `QFile` internally does) may work for getting file descriptor.

lukas-w replied:
Neither is really better than the other, but they do the same thing so I think they should be consolidated into a single function. The Qt solution is not an option for this because it's not available in `RemoteVstPlugin`.

PhysSong replied:
> but they do the same thing
They *almost* do. On Windows, `QFile` holds native file handle and converts it to a C file handle when requested.
`_wfopen` creates a `FILE` which holds a C handle to the file. That's why I used the former here, but I can switch to the latter if needed.

lukas-w replied:
Surely `F_OPEN` and `fdopenFromUtf8` fulfil the same task, which is opening a C `FILE` given an UTF8-encoded file name, right? Even though they don't work the same way internally, they "do the same thing" as in "they solve the same problem". From that point of view, this is duplicate code and there should only be one function for it.

PhysSong replied:
Then where should I move `F_OPEN` and related stuffs? I can't find a good place for it.
I was not using that in Zyn because it(more precisely `zlib`) needs `fdopen`-able handle, not a `FILE` pointer. However, getting(and duplicating) file handle is not hard so I can easily add it.

lukas-w replied:
Good question, I can't find a good place either. Maybe just create a new header? Something like `lmmsio.h` (inspired by `stdio.h`), although that name is very similar to https://lmms.io/.

PhysSong replied:
I also used `QFile` for getting file descriptors which will be used by `libsndfile`. Do you think it's worth unifying everything with `F_OPEN` based macros/functions?

PhysSong replied:
I think either `lmms_io.h` or `IoHelper.h` will be fine. I'll address this point soon.

PhysSong replied:
Or how about putting them in `RemotePlugin.h`? I can't use-cases of UTF-8 string filename except in remote plugins.

PhysSong replied:
Is it fine for our Zyn to include LMMS' headers at all? I think this will be fine in `stable-1.2`, but I'm not sure about `master`'s submodule. It would be the last concern for me...

PhysSong replied:
I know it doesn't make sense to build https://github.com/LMMS/zynaddsubfx as a standalone, of course. However, I think it's possible to check if it's building with LMMS or not, though not really needed.

PhysSong commented:
@lukas-w I addressed these issues and pushed to a [separate branch](https://github.com/PhysSong/lmms/tree/encoding2). I'll push this change to the original branch when/if you says okay.

PhysSong commented:
Oops, accidentally pushed them. I can roll back to [original commits](https://github.com/PhysSong/lmms/commit/1f09c05133722a5274782bb0d1d1078f200a3f80) whenever needed.

lukas-w commented:
@PhysSong Please roll back to the original commits, all the comment/review history on GitHub is lost otherwise. It's okay to have one extra commit on top of the original ones, there's no need to rewrite history.

PhysSong commented:
> all the comment/review history on GitHub is lost otherwise.

What do you mean? Review comments are still there unless you commented on specific commits, the are just collapsed.

Anyway, I agree that it's not necessary to rewrite history. I just wanted to keep things cleaner, but that was all. If you're still unhappy with this force-push, I can roll back.

lukas-w commented:
> What do you mean?

Sorry, I was being imprecise. The comments are still there, but the context is lost. Review comments are always on specific commits and you can't view that commits any more, so you can't see the changes the review refers to. Try clicking on the "View changes" button next to a review, this is what you'll get now:

![image](https://user-images.githubusercontent.com/2879917/41977744-6c1236a8-7a20-11e8-9129-86e9457c0f16.png)

> If you're still unhappy with this force-push, I can roll back.

Yes, please do 👍

PhysSong commented:
I just noticed even a rebasing can cause such a thing. I just accidentally rebased commits because I didn't know that... I'll revert to the revision before rebasing if needed.

lukas-w commented:
@PhysSong Apart from restoring the pre-force-push commits, nope.

PhysSong commented:
Then do you mind if I merge this(probably with cleaner history)?

lukas-w commented:
I don't mind, go ahead ❤️

lukas-w commented on plugins/vst_base/RemoteVstPlugin.cpp:
```diff
@@ -1072,13 +1073,13 @@ void RemoteVstPlugin::saveChunkToFile( const std::string & _file )
-			int fd = open( _file.c_str(), O_WRONLY | O_BINARY );
-			if ( ::write( fd, chunk, len ) != len )
+			FILE* fp = F_OPEN_UTF8( _file, "wb" );
+			if ( fwrite( chunk, len, 1, fp ) != len )
```
Only spotted this now: `fwrite`returns the `count` of elements successfully written, so in this case it will return `1` on success, not `len`. Therefore this condition is always leading to a false error message on success (except when coincidentally `len == 1`).

PhysSong replied:
Thank you! Changed to `fwrite( chunk, 1, len, fp )` in 0f3b41f590ed8d573ce3891f6502f211d1be176a.

PhysSong replied:
Also synced into `master` via 170a46e7a3b5bc3eb49999874344b82cda1cc53f.
